crates/fuzzing/src/generators/single_inst_module.rs

```diff
@@ -0,0 +1,349 @@
+//! Generate Wasm modules that contain a single instruction.
+
+use arbitrary::{Arbitrary, Unstructured};
+use once_cell::sync::Lazy;
+use wasm_encoder::{
+    CodeSection, ExportKind, ExportSection, Function, FunctionSection, Instruction, Module,
+    TypeSection, ValType,
+};
+
+/// The name of the function generated by this module.
+const FUNCTION_NAME: &'static str = "test";
+
+/// Configure a single instruction module.
+///
+/// By explicitly defining the parameter and result types (versus generating the
+/// module directly), we can more easily generate values of the right type.
+#[derive(Clone, Debug)]
+pub struct SingleInstModule<'a> {
+    instruction: Instruction<'a>,
+    parameters: Vec<ValType>,
+    results: Vec<ValType>,
+}
+
+impl<'a> SingleInstModule<'a> {
+    /// Generate a binary Wasm module with a single exported function, `test`,
+    /// that executes the single instruction.
+    pub fn encode(&self) -> Vec<u8> {
+        let mut module = Module::new();
+
+        // Encode the type section.
+        let mut types = TypeSection::new();
+        types.function(self.parameters.clone(), self.results.clone());
+        module.section(&types);
+
+        // Encode the function section.
+        let mut functions = FunctionSection::new();
+        let type_index = 0;
+        functions.function(type_index);
+        module.section(&functions);
+
+        // Encode the export section.
+        let mut exports = ExportSection::new();
+        exports.export(FUNCTION_NAME, ExportKind::Func, 0);
+        module.section(&exports);
+
+        // Encode the code section.
+        let mut codes = CodeSection::new();
+        let locals = vec![];
+        let mut f = Function::new(locals);
+        for (index, _) in self.parameters.iter().enumerate() {
+            f.instruction(&Instruction::LocalGet(index as u32));
+        }
+        f.instruction(&self.instruction);
+        f.instruction(&Instruction::End);
+        codes.function(&f);
+        module.section(&codes);
+
+        // Extract the encoded Wasm bytes for this module.
+        module.finish()
+    }
+}
+
+impl<'a> Arbitrary<'a> for SingleInstModule<'_> {
+    fn arbitrary(u: &mut Unstructured<'a>) -> arbitrary::Result<Self> {
+        u.choose(&INSTRUCTIONS).cloned()
```

Comment:
Does the module need to be cloned, or can this be implemented for `&SingleInstModule<'_>` instead?